Promtail migration guide (#4709)
grafana/agent: docs/sources/flow/getting-started/migrating-from-promtail.md

   from `OUTPUT_CONFIG_PATH`:

### Debugging

1. If the convert command cannot convert a Promtail configuration, diagnostic
   information is sent to `stderr`. You can bypass any non-critical issues and
   output the flow configuration using a best-effort conversion by including
   the `--bypass-errors` flag.

   {{% admonition type="caution" %}}If you bypass the errors, the behavior of the converted configuration may not match the original Promtail configuration. Make sure you fully test the converted configuration before using it in a production environment.{{% /admonition %}}

    ```bash
    AGENT_MODE=flow; grafana-agent convert --source-format=promtail --bypass-errors --output=OUTPUT_CONFIG_PATH INPUT_CONFIG_PATH
    ```

1. You can also output a diagnostic report by including the `--report` flag.

    ```bash
    AGENT_MODE=flow; grafana-agent convert --source-format=promtail --report=OUTPUT_REPORT_PATH --output=OUTPUT_CONFIG_PATH INPUT_CONFIG_PATH
    ```

    * Replace `OUTPUT_REPORT_PATH` with the output path for the report.

   Using the [example](#example) Promtail configuration below, the diagnostic
   report provides the following information:

    ```plaintext

>     (Warning) If you have a tracing set up for Promtail, it cannot be migrated to Flow Mode automatically - please refer to documentation on how to configure tracing in Flow Mode

Reviewer: ```suggestion
    (Warning) If you have a tracing set up for Promtail, it cannot be migrated to Flow Mode automatically. Refer to the documentation on how to configure tracing in Flow Mode.
```

I've also suggested the same change in the go files.

1. Generally we should avoid using `please` in technical documentation. It's not wrong, but it is also not necessary. Keep things simple and concise. (also, using `please` implies we are asking the user for a favour vs instructing them on the task).
1. Does the user know where this reference documentation is located?

Author: Re: 2. They should ideally be able to find this with ease in our docs. Hard-coding a link in the binary may not be the best idea, but we could consider putting it in some variable somewhere.

>     (Warning) The Agent Flow Mode's metrics are different from the metrics emitted by Promtail. If you rely on Promtail's metrics you will need to change your configuration (e.g. alerts and dashboards).

Reviewer: Same here. I updated the go files as well.

```suggestion
    (Warning) The Agent Flow Mode's metrics are different from the metrics emitted by Promtail. If you rely on Promtail's metrics, you must change your configuration, for example, your alerts and dashboards.
```

    ```

## Run a Promtail configuration

If you’re not ready to completely switch to a flow configuration, you can run
Grafana Agent using your existing Promtail configuration.
The `--config.format=promtail` flag tells Grafana Agent to convert your Promtail
configuration to Flow Mode and load it directly without saving the new
configuration. This allows you to try Flow Mode without modifying your existing
Promtail configuration infrastructure.

> In this task, we will use the [run][] CLI command to run Grafana Agent in Flow
> Mode using a Promtail configuration.

> [Start the agent][] in Flow Mode and include the command line flag

Reviewer: ```suggestion
[Start the Agent][] in flow mode and include the command line flag
```

`--config.format=promtail`. Your configuration file must be a valid Promtail
configuration file rather than a Flow Mode configuration file.

### Debugging

1. You can follow the convert CLI command [debugging][] instructions to generate
   a diagnostic report.

> 1. Refer to the Grafana Agent [Flow Debugging][] for more information about a

Reviewer: ```suggestion
1. Refer to the Grafana Agent [Flow Debugging][] for more information about
```

## Example

This example demonstrates converting a Promtail configuration file to a Grafana
Agent Flow Mode configuration file.

The following Promtail configuration file provides the input for the conversion:

```yaml
clients:
  - url: http://localhost/loki/api/v1/push
scrape_configs:
  - job_name: example
    static_configs:
      - targets:
          - localhost
        labels:
          __path__: /var/log/*.log
```

The convert command takes the YAML file as input and outputs a River file.

```bash
AGENT_MODE=flow; grafana-agent convert --source-format=promtail --output=OUTPUT_CONFIG_PATH INPUT_CONFIG_PATH
```

The new Flow Mode configuration file looks like this:

```river
local.file_match "example" {
	path_targets = [{
		__address__ = "localhost",
		__path__    = "/var/log/*.log",
	}]
}

loki.source.file "example" {
	targets    = local.file_match.example.targets
	forward_to = [loki.write.default.receiver]
}

loki.write "default" {
	endpoint {
		url = "http://localhost/loki/api/v1/push"
	}
	external_labels = {}
}
```

## Limitations

Config conversion is done on the best-effort basis; the Agent will issue
warnings or errors where the conversion cannot be performed.

> Once the configuration is converted, it is recommended for users to understand
> and review the output Flow Mode configuration and verify that it is correct
> before starting to use it in a production environment.

Reviewer: ```suggestion
Once the configuration is converted, we recommend that you review
the Flow Mode configuration file created, and verify that it is correct
before starting to use it in a production environment.
```

> Furthermore, it is recommended to go through the following checklist:

Reviewer: ```suggestion
Furthermore, we recommend that you review the following checklist:
```

* Check if you are using any extra command line arguments with Promtail which
  are not present in your config file, e.g. `-max-line-size`
* Check if you are setting any environment variables -
  whether [expanded in the config file][] itself or consumed directly by
  Promtail such as `JAEGER_AGENT_HOST`.
* In Flow Mode, the positions file will be saved at a different location -
  see [loki.source.file][] for more details. Check if you have any existing
  setup (e.g. Kubernetes Persistent Volume) that needs changing to use the new
  positions file path.
* Metrics exposed by the Flow Mode will usually match those of Promtail, but
  will use a different name. Make sure that you use the new metric names, e.g.
  in your alerts and dashboards queries.
* Please note that the logs produced by the Agent will differ from those
  produced by Promtail.
* Please note that the Agent exposes [Grafana Agent Flow UI][], which differs
  from Promtail's Web UI.

[expanded in the config file]: https://grafana.com/docs/loki/latest/clients/promtail/configuration/#use-environment-variables-in-the-configuration

> [Grafana Agent Flow UI]: http://localhost:3003/docs/agent/latest/flow/monitoring/debugging/#grafana-agent-flow-ui

Reviewer 2: check the url (localhost)

Author: Whoops, sorry! Was a bit in a rush when I pushed this. Will clean up now.
